design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+# KLIP 47 - NON NULL columns and component types
+
+**Author**: Steven Zhang (@stevenpyzhang) | 
+**Release Target**: TBD | 
+**Status**: _In Discussion_ | 
+**Discussion**: TBD-PR-LINK
+
+**tl;dr:** _NOT NULL increases data integrity by putting constraints on column values in streams and tables._
+           
+## Motivation and background
+https://github.com/confluentinc/ksql/issues/4436
+
+The SQL standard allows a column to have a `NOT NULL` constraint. `NOT NULL` indicates that
+`NULL` is not an accepted value for that column. This constraint prevents records with a `NULL` column value from being written into the table. 
+When a user reads from the table, its guaranteed that every table record will have a value in the `NOT NULL` column.
+
+Currently, all columns in KSQL are implicitly nullable. This KLIP would introduce `NOT NULL` syntax so users can apply it as a constraint to ksqlDB columns.
+Columns that don't have `NOT NULL` constraint default to being nullable, which is the current behavior. When a `NOT NULL` constraint is applied to a column, 
+records with null column values should not be written into constrained stream/table via `CSAS/CTAS/Insert Into` queries or `Insert Into Values` statements.
```

Comment:
I don't think we should include `CSAS/CTAS` as the result of a query cannot be constraints IMHO.

If people want to ensure that no `NULL` values are in the result, they need to specify corresponding filters via `WHERE` or `HAVING` to drop columns that have a `NULL` value.

Reply:
Just my 2c: 

1) if we do not consider UDF/UDAF for now, then `NOT NULL` constraint should always be "inherited" in the derived table/streams since no built-in operators so far would map a not-null field into a null field in the derived entities.

2) in addition, for built-in operators, I think they should be able to auto handling null fields, so users should not need to worry if they need to explicitly specify "I want to omit those null-fielded records in the expression/etc". @stevenpyzhang please correct me if I'm wrong.

3) for built-in UDF/UDAF, I think user's should be responsible to make sure that their implementation can handle null fields, and decide whether the output could be null or not-null.

With all that, I feel we do not need to support adding constraint on the CT/SAS statements, since either it should be enforced at the complied code level, or it should not be required. Thoughts?

Reply:
> constraint should always be "inherited" in the derived table/streams since no built-in operators so far would map a not-null field into a null field in the derived entities

I don't think this hold true. For example, left/outer joins may produce `NULL` even if no input stream/table contains any `NULL`s.

However, I don't think this is the point I try to make: If you define a _materialized view_, you define a "dependency" from the inputs to the output and this dependency is described by the _semantics of the query_: **The output stream/table must always contain all rows that are the result of the query.** Or to rephrase: if you do a (transient) query you define a query that gives you a result that may or may not contain `NULL`s. The `CMV` prefix **only** tells the system to _store_ the result (and to also maintain it to be "in-sync" if the input changes); the `CMV` statement must **not modify** the result by any means IMHO. Thus, if you have for example a left outer join, and the result contains `NULL`, the result _MUST_ contain this row. Otherwise the contract of the MV definition is violated. -- To me, it seems to be _fundamentally_ incorrect to define a constraint on a _query result_...

Reply:
> constraint should always be "inherited" in the derived table/streams since no built-in operators so far would map a not-null field into a null field in the derived entities

I don't think this hold true. For example, left/outer joins may produce `NULL` even if no input stream/table contains any `NULL`s.

However, I don't think this is the point I try to make: If you define a _materialized view_, you define a "dependency" from the inputs to the output and this dependency is described by the _semantics of the query_: **The output stream/table must always contain all rows that are the result of the query.** Or to rephrase: if you do a (transient) query you define a query that gives you a result that may or may not contain `NULL`s. The `CMV` prefix **only** tells the system to _store_ the result (and to also maintain it to be "in-sync" if the input changes); the `CMV` statement must **not modify** the result by any means IMHO. Thus, if you have for example a left outer join, and the result contains `NULL`, the result _MUST_ contain this row. Otherwise the contract of the MV definition is violated. -- To me, it seems to be _fundamentally_ incorrect to define a constraint on a _query result_...

Reply:
@mjsax @guozhangwang if I understand correctly, you both are saying, that constraints shouldn't be set on derived streams/tables directly, but rather infered from the constraints of the source.
Given the following statement:
```sql
CREATE STREAM target 
AS SELECT col1, col2 
FROM source
WHERE col2 IS NOT NULL;
```
If col1 from source is not null, then col1 in target shouldn't be nullable as well and col2 in target is never nullable. Or:
```sql
CREATE STREAM target 
AS SELECT COALESCE(col1, col2) AS col3 
FROM source;
```
If col2 from source is not null, then col3 in target shouldn't be nullable. This means any function should declare whether it produces nullable or not nullable results (and under which circumstances).
Also some clauses may affect nullability, eg. inner join forces not null for join criteria, left/outer join forces nullable.

In general it should be possible to derive the constraints of the target stream/table from the constraints of the source and the statement which creates it. Would you agree wirh that?